omaha/core/scheduler.cc

```diff
@@ -14,116 +14,130 @@
 // ========================================================================
 
 #include "omaha/core/scheduler.h"
-#include <algorithm>
+
 #include "omaha/base/debug.h"
 #include "omaha/base/error.h"
-#include "omaha/base/highres_timer-win32.h"
 #include "omaha/base/logging.h"
-#include "omaha/base/queue_timer.h"
-#include "omaha/common/config_manager.h"
-#include "omaha/core/core.h"
-#include "omaha/core/core_metrics.h"
 
 namespace omaha {
 
-Scheduler::Scheduler(const Core& core)
-    : core_(core) {
-  CORE_LOG(L1, (_T("[Scheduler::Scheduler]")));
+Scheduler::SchedulerItem::SchedulerItem(HANDLE timer_queue,
+                                        int start_delay_ms,
+                                        int interval_ms,
+                                        bool has_debug_timer,
+                                        ScheduledWork work)
+    : start_delay_ms_(start_delay_ms), interval_ms_(interval_ms), work_(work) {
+  if (has_debug_timer) {
+    debug_timer_.reset(new HighresTimer());
+  }
+
+  if (timer_queue) {
+    timer_.reset(
+        new QueueTimer(timer_queue, &SchedulerItem::TimerCallback, this));
+    VERIFY1(SUCCEEDED(
+        ScheduleNext(timer_.get(), debug_timer_.get(), start_delay_ms)));
+  }
 }
 
-Scheduler::~Scheduler() {
-  CORE_LOG(L1, (_T("[Scheduler::~Scheduler]")));
+Scheduler::SchedulerItem::~SchedulerItem() {
+  // QueueTimer dtor may block for pending callbacks
+  timer_.reset(nullptr);
```

Comment:
Can we use the same syntax to reset the pointers? I suggest we be consistent, for example, in this assignment to nullptr is used too.

Reply:
I think I made it confusing by assigning `nullptr` to `timer_queue_` which isn't a smart pointer. It's just a handle, so now I assign it to `NULL`. The rest uses the reset syntax.